File: app/src/dataSource/dataPipeline.ts
import type { IRow, IDataQueryPayload } from '@kanaries/graphic-walker/interfaces';
import type { IFilterCondition, FilterLogic } from '../interfaces/filter';

export enum DataPipelineStage {
    SOURCE = "source",
    FILTER = "filter",
    TRANSFORM = "transform",
    SAMPLE = "sample",
    AGGREGATE = "aggregate",
    OUTPUT = "output"
}

export interface PipelineContext {
    sourceType: 'client' | 'server';
    payload?: IDataQueryPayload;
    sql?: string;
    filters: IFilterCondition[];
    filterLogic: FilterLogic;
    samplingStrategy?: 'smart' | 'random' | 'byte_limit' | 'scatter';
    sampleSize?: number;
    byteLimit?: number;
    preserveBoundaries: boolean;
    metadata: Record<string, any>;
}

export interface DataSource {
    getDatasByPayload(payload: IDataQueryPayload): Promise<IRow[]>;
    getDatasBySql(sql: string): Promise<IRow[]>;
    batchGetDatasByPayload(payloadList: IDataQueryPayload[]): Promise<IRow[][]>;
    batchGetDatasBySql(sqlList: string[]): Promise<IRow[][]>;
}

export interface LocalDataSource {
    data: IRow[];
}

export type PipelineExecutor = 
    | { type: 'client'; source: LocalDataSource }
    | { type: 'server'; source: DataSource };

export class PipelineStage {
    name: string;
    
    constructor(name: string) {
        this.name = name;
    }
    
    execute(data: IRow[], context: PipelineContext): IRow[] {
        return data;
    }
}

export class FilterStage extends PipelineStage {
    constructor() {
        super('filter');
    }
    
    execute(data: IRow[], context: PipelineContext): IRow[] {
        if (!context.filters || context.filters.length === 0) {
            return data;
        }
        
        const enabledFilters = context.filters.filter(c => c.enabled);
        if (enabledFilters.length === 0) {
            return data;
        }
        
        return data.filter(row => {
            if (context.filterLogic === "AND") {
                return enabledFilters.every(filter => this.matchCondition(row, filter));
            } else {
                return enabledFilters.some(filter => this.matchCondition(row, filter));
            }
        });
    }
    
    private matchCondition(row: IRow, condition: IFilterCondition): boolean {
        const value = row[condition.fid];
        
        switch (condition.conditionType) {
            case 'range':
                return this.matchRange(value, condition.value);
            case 'temporal range':
                return this.matchTemporalRange(value, condition.value);
            case 'one of':
                return this.matchOneOf(value, condition.value);
            case 'not in':
                return !this.matchOneOf(value, condition.value);
            case 'contains':
                return this.matchContains(value, condition.value);
            case 'equals':
                return value === condition.value;
            case 'greater than':
                return this.matchGreaterThan(value, condition.value);
            case 'less than':
                return this.matchLessThan(value, condition.value);
            default:
                return true;
        }
    }
    
    private matchRange(value: any, rangeValue: [number | null, number | null]): boolean {
        if (!Array.isArray(rangeValue) || rangeValue.length !== 2) {
            return true;
        }
        const [min, max] = rangeValue;
        if (typeof value !== 'number') return false;
        const aboveMin = min === null || value >= min;
        const belowMax = max === null || value <= max;
        return aboveMin && belowMax;
    }
    
    private matchTemporalRange(value: any, rangeValue: [number | null, number | null]): boolean {
        if (!Array.isArray(rangeValue) || rangeValue.length !== 2) {
            return true;
        }
        const [min, max] = rangeValue;
        
        let timestamp: number;
        if (value instanceof Date) {
            timestamp = value.getTime();
        } else if (typeof value === 'string') {
            timestamp = new Date(value).getTime();
        } else if (typeof value === 'number') {
            timestamp = value;
        } else {
            return false;
        }
        
        if (isNaN(timestamp)) return false;
        
        const aboveMin = min === null || timestamp >= min;
        const belowMax = max === null || timestamp <= max;
        return aboveMin && belowMax;
    }
    
    private matchOneOf(value: any, values: any[]): boolean {
        if (!Array.isArray(values)) {
            return true;
        }
        return values.includes(value);
    }
    
    private matchContains(value: any, substring: string): boolean {
        if (typeof value !== 'string' || typeof substring !== 'string') {
            return false;
        }
        return value.toLowerCase().includes(substring.toLowerCase());
    }
    
    private matchGreaterThan(value: any, threshold: number): boolean {
        if (threshold === null || threshold === undefined) {
            return true;
        }
        if (typeof value !== 'number' || typeof threshold !== 'number') {
            return false;
        }
        return value > threshold;
    }
    
    private matchLessThan(value: any, threshold: number): boolean {
        if (threshold === null || threshold === undefined) {
            return true;
        }
        if (typeof value !== 'number' || typeof threshold !== 'number') {
            return false;
        }
        return value < threshold;
    }
}

export class SamplingStage extends PipelineStage {
    constructor() {
        super('sample');
    }
    
    execute(data: IRow[], context: PipelineContext): IRow[] {
        if (!context.samplingStrategy === undefined) {
            return data;
        }
        
        switch (context.samplingStrategy) {
            case 'smart':
                return this.smartSample(data, context);
            case 'random':
                return this.randomSample(data, context);
            case 'scatter':
                return this.scatterSample(data, context);
            default:
                return data;
        }
    }
    
    private smartSample(data: IRow[], context: PipelineContext): IRow[] {
        if (context.sampleSize === undefined || data.length <= context.sampleSize) {
            return data;
        }
        
        const n = data.length;
        const sampleSize = context.sampleSize;
        
        if (context.preserveBoundaries && sampleSize >= 2) {
            const boundaryCount = Math.min(2, sampleSize);
            const middleCount = sampleSize - boundaryCount;
            const middleData = boundaryCount === 2 
                ? data.slice(boundaryCount, -boundaryCount) 
                : data.slice(1);
            
            let sampled: IRow[];
            if (middleData.length <= middleCount) {
                sampled = middleData;
            } else {
                const step = middleData.length / middleCount;
                sampled = [];
                for (let i = 0; i < middleCount; i++) {
                    sampled.push(middleData[Math.floor(i * step)]);
                }
            }
            
            if (boundaryCount === 2) {
                return [data[0], ...sampled, data[data.length - 1]];
            } else {
                return [data[0], ...sampled];
            }
        }
        
        const step = n / sampleSize;
        const result: IRow[] = [];
        for (let i = 0; i < sampleSize; i++) {
            result.push(data[Math.floor(i * step)]);
        }
        return result;
    }
    
    private randomSample(data: IRow[], context: PipelineContext): IRow[] {
        if (context.sampleSize === undefined || data.length <= context.sampleSize) {
            return data;
        }
        
        const indices = new Set<number>();
        while (indices.size < context.sampleSize) {
            indices.add(Math.floor(Math.random() * data.length));
        }
        return Array.from(indices).sort((a, b) => a - b).map(i => data[i]);
    }
    
    private scatterSample(data: IRow[], context: PipelineContext): IRow[] {
        const SCATTER_PLOT_LARGE_DATA_THRESHOLD = 10000;
        const SCATTER_PLOT_SAMPLE_LIMIT = 5000;
        
        const n = data.length;
        
        if (n <= SCATTER_PLOT_LARGE_DATA_THRESHOLD) {
            return data;
        }
        
        const sampleSize = Math.min(SCATTER_PLOT_SAMPLE_LIMIT, Math.floor(SCATTER_PLOT_LARGE_DATA_THRESHOLD / 2));
        
        return this.smartSample(data, {
            ...context,
            sampleSize,
            preserveBoundaries: true
        });
    }
}

export class DataPipeline {
    executor: PipelineExecutor;
    stages: PipelineStage[];
    context: PipelineContext;
    
    constructor(
        executor: PipelineExecutor,
        stages?: PipelineStage[],
        context?: Partial<PipelineContext>
    ) {
        this.executor = executor;
        this.stages = stages || [new FilterStage(), new SamplingStage()];
        this.context = {
            sourceType: executor.type,
            filters: [],
            filterLogic: 'AND',
            preserveBoundaries: false,
            metadata: {},
            ...context
        };
    }
    
    static create(
        executor: PipelineExecutor,
        context?: Partial<PipelineContext>
    ): DataPipeline {
        return new DataPipeline(executor, undefined, context);
    }
    
    configure(updates: Partial<PipelineContext>): DataPipeline {
        this.context = { ...this.context, ...updates };
        return this;
    }
    
    addStage(stage: PipelineStage): DataPipeline {
        this.stages.push(stage);
        return this;
    }
    
    private async fetchData(): Promise<IRow[]> {
        if (this.executor.type === 'client') {
            return [...this.executor.source.data];
        }
        
        if (this.context.payload !== undefined) {
            return this.executor.source.getDatasByPayload(this.context.payload);
        }
        
        if (this.context.sql !== undefined) {
            return this.executor.source.getDatasBySql(this.context.sql);
        }
        
        throw new Error('Either payload or sql must be provided in context');
    }
    
    async execute(
        payload?: IDataQueryPayload,
        sql?: string,
        updates?: Partial<PipelineContext>
    ): Promise<IRow[]> {
        if (payload !== undefined) {
            this.context.payload = payload;
        }
        if (sql !== undefined) {
            this.context.sql = sql;
        }
        if (updates !== undefined) {
            this.configure(updates);
        }
        
        let data = await this.fetchData();
        
        for (const stage of this.stages) {
            data = stage.execute(data, this.context);
        }
        
        return data;
    }
    
    async executeBatch(
        payloadList?: IDataQueryPayload[],
        sqlList?: string[],
        updates?: Partial<PipelineContext>
    ): Promise<IRow[][]> {
        if (updates !== undefined) {
            this.configure(updates);
        }
        
        let rawDatas: IRow[][];
        const executor = this.executor;
        
        if (executor.type === 'client') {
            rawDatas = payloadList?.map(() => [...executor.source.data]) ?? [];
        } else if (payloadList !== undefined) {
            rawDatas = await executor.source.batchGetDatasByPayload(payloadList);
        } else if (sqlList !== undefined) {
            rawDatas = await executor.source.batchGetDatasBySql(sqlList);
        } else {
            throw new Error('Either payloadList or sqlList must be provided');
        }
        
        const results: IRow[][] = [];
        for (const data of rawDatas) {
            let processedData = data;
            for (const stage of this.stages) {
                processedData = stage.execute(processedData, this.context);
            }
            results.push(processedData);
        }
        
        return results;
    }
}

export class PipelineBuilder {
    executor: PipelineExecutor;
    stages: PipelineStage[] = [];
    contextUpdates: Partial<PipelineContext> = {};
    
    constructor(executor: PipelineExecutor) {
        this.executor = executor;
    }
    
    withFilter(filters: IFilterCondition[], logic: FilterLogic = 'AND'): PipelineBuilder {
        if (filters.length > 0) {
            this.stages.push(new FilterStage());
            this.contextUpdates.filters = filters;
            this.contextUpdates.filterLogic = logic;
        }
        return this;
    }
    
    withSampling(
        strategy: 'smart' | 'random' | 'scatter',
        sampleSize?: number,
        preserveBoundaries: boolean = false
    ): PipelineBuilder {
        this.stages.push(new SamplingStage());
        this.contextUpdates.samplingStrategy = strategy;
        if (sampleSize !== undefined) {
            this.contextUpdates.sampleSize = sampleSize;
        }
        this.contextUpdates.preserveBoundaries = preserveBoundaries;
        return this;
    }
    
    withScatterSampling(): PipelineBuilder {
        return this.withSampling('scatter');
    }
    
    withSmartSampling(sampleSize: number, preserveBoundaries: boolean = true): PipelineBuilder {
        return this.withSampling('smart', sampleSize, preserveBoundaries);
    }
    
    withMetadata(metadata: Record<string, any>): PipelineBuilder {
        this.contextUpdates.metadata = {
            ...this.contextUpdates.metadata,
            ...metadata
        };
        return this;
    }
    
    build(): DataPipeline {
        return new DataPipeline(
            this.executor,
            this.stages,
            this.contextUpdates
        );
    }
}

export function createPipelineBuilder(executor: PipelineExecutor): PipelineBuilder {
    return new PipelineBuilder(executor);
}

export function createClientPipeline(
    data: IRow[],
    filters?: IFilterCondition[],
    filterLogic?: FilterLogic
): DataPipeline {
    const builder = createPipelineBuilder({
        type: 'client',
        source: { data }
    });
    
    if (filters && filters.length > 0) {
        builder.withFilter(filters, filterLogic);
    }
    
    return builder.build();
}

export function createServerPipeline(
    source: DataSource,
    filters?: IFilterCondition[],
    filterLogic?: FilterLogic
): DataPipeline {
    const builder = createPipelineBuilder({
        type: 'server',
        source
    });
    
    if (filters && filters.length > 0) {
        builder.withFilter(filters, filterLogic);
    }
    
    return builder.build();
}

File: pygwalker/services/data_pipeline.py
from typing import List, Dict, Any, Optional, Union, Callable, Protocol, runtime_checkable
from dataclasses import dataclass, field
from enum import Enum
import logging

logger = logging.getLogger(__name__)


class DataPipelineStage(Enum):
    """数据管道阶段枚举"""
    SOURCE = "source"
    FILTER = "filter"
    TRANSFORM = "transform"
    SAMPLE = "sample"
    AGGREGATE = "aggregate"
    OUTPUT = "output"


@runtime_checkable
class DataSource(Protocol):
    """数据源协议 - 定义数据获取接口"""
    
    def get_datas_by_payload(self, payload: Dict[str, Any]) -> List[Dict[str, Any]]:
        """通过 payload 获取数据"""
        ...
    
    def get_datas_by_sql(self, sql: str) -> List[Dict[str, Any]]:
        """通过 SQL 获取数据"""
        ...
    
    def batch_get_datas_by_payload(self, payload_list: List[Dict[str, Any]]) -> List[List[Dict[str, Any]]]:
        """批量通过 payload 获取数据"""
        ...
    
    def batch_get_datas_by_sql(self, sql_list: List[str]) -> List[List[Dict[str, Any]]]:
        """批量通过 SQL 获取数据"""
        ...


@dataclass
class PipelineContext:
    """数据管道上下文 - 存储管道执行过程中的状态和元数据"""
    
    source_type: str
    payload: Optional[Dict[str, Any]] = None
    sql: Optional[str] = None
    filters: List[Dict[str, Any]] = field(default_factory=list)
    filter_logic: str = "AND"
    sampling_strategy: Optional[str] = None
    sample_size: Optional[int] = None
    byte_limit: Optional[int] = None
    preserve_boundaries: bool = False
    metadata: Dict[str, Any] = field(default_factory=dict)
    
    def update(self, **kwargs) -> "PipelineContext":
        """更新上下文属性"""
        for key, value in kwargs.items():
            setattr(self, key, value)
        return self


class PipelineStage:
    """管道阶段基类"""
    
    def __init__(self, name: str):
        self.name = name
    
    def execute(self, data: List[Dict[str, Any]], context: PipelineContext) -> List[Dict[str, Any]]:
        """执行阶段 - 子类需要重写此方法"""
        raise NotImplementedError
    
    def __call__(self, data: List[Dict[str, Any]], context: PipelineContext) -> List[Dict[str, Any]]:
        return self.execute(data, context)


class FilterStage(PipelineStage):
    """过滤阶段 - 应用过滤条件到数据"""
    
    def __init__(self):
        super().__init__("filter")
    
    def execute(self, data: List[Dict[str, Any]], context: PipelineContext) -> List[Dict[str, Any]]:
        if not context.filters:
            return data
        
        enabled_filters = [f for f in context.filters if f.get("enabled", True)]
        if not enabled_filters:
            return data
        
        try:
            return self._apply_filters(data, enabled_filters, context.filter_logic)
        except Exception as e:
            logger.warning(f"Filter stage failed: {e}, returning original data")
            return data
    
    def _apply_filters(
        self, 
        data: List[Dict[str, Any]], 
        filters: List[Dict[str, Any]], 
        logic: str
    ) -> List[Dict[str, Any]]:
        """应用多个过滤条件"""
        if logic == "AND":
            return [
                row for row in data 
                if all(self._match_condition(row, f) for f in filters)
            ]
        else:
            return [
                row for row in data 
                if any(self._match_condition(row, f) for f in filters)
            ]
    
    def _match_condition(self, row: Dict[str, Any], condition: Dict[str, Any]) -> bool:
        """检查单行是否满足过滤条件"""
        fid = condition.get("fid")
        condition_type = condition.get("conditionType", condition.get("type"))
        value = condition.get("value")
        
        if fid not in row:
            return False
        
        row_value = row[fid]
        
        if condition_type == "range":
            return self._match_range(row_value, value)
        elif condition_type == "temporal range":
            return self._match_temporal_range(row_value, value)
        elif condition_type == "one of":
            return self._match_one_of(row_value, value)
        elif condition_type == "not in":
            return not self._match_one_of(row_value, value)
        elif condition_type == "contains":
            return self._match_contains(row_value, value)
        elif condition_type == "equals":
            return row_value == value
        elif condition_type == "greater than":
            return self._match_greater_than(row_value, value)
        elif condition_type == "less than":
            return self._match_less_than(row_value, value)
        elif condition.get("rule"):
            rule = condition["rule"]
            rule_type = rule.get("type")
            rule_value = rule.get("value")
            if rule_type == "range":
                return self._match_range(row_value, rule_value)
            elif rule_type == "temporal range":
                return self._match_temporal_range(row_value, rule_value)
            elif rule_type == "one of":
                return self._match_one_of(row_value, rule_value)
            elif rule_type == "not in":
                return not self._match_one_of(row_value, rule_value)
        
        return True
    
    def _match_range(self, value: Any, range_value: List[Any]) -> bool:
        if not isinstance(range_value, (list, tuple)) or len(range_value) != 2:
            return True
        min_val, max_val = range_value
        if not isinstance(value, (int, float)):
            return False
        above_min = min_val is None or value >= min_val
        below_max = max_val is None or value <= max_val
        return above_min and below_max
    
    def _match_temporal_range(self, value: Any, range_value: List[Any]) -> bool:
        if not isinstance(range_value, (list, tuple)) or len(range_value) != 2:
            return True
        min_val, max_val = range_value
        
        import datetime
        try:
            if isinstance(value, datetime.datetime):
                timestamp = value.timestamp() * 1000
            elif isinstance(value, str):
                timestamp = datetime.datetime.fromisoformat(value.replace('Z', '+00:00')).timestamp() * 1000
            else:
                timestamp = float(value)
            
            above_min = min_val is None or timestamp >= min_val
            below_max = max_val is None or timestamp <= max_val
            return above_min and below_max
        except (ValueError, TypeError):
            return False
    
    def _match_one_of(self, value: Any, values: List[Any]) -> bool:
        if not isinstance(values, (list, tuple, set)):
            return True
        return value in values
    
    def _match_contains(self, value: Any, substring: str) -> bool:
        if not isinstance(value, str) or not isinstance(substring, str):
            return False
        return substring.lower() in value.lower()
    
    def _match_greater_than(self, value: Any, threshold: Any) -> bool:
        if threshold is None:
            return True
        if not isinstance(value, (int, float)) or not isinstance(threshold, (int, float)):
            return False
        return value > threshold
    
    def _match_less_than(self, value: Any, threshold: Any) -> bool:
        if threshold is None:
            return True
        if not isinstance(value, (int, float)) or not isinstance(threshold, (int, float)):
            return False
        return value < threshold


class SamplingStage(PipelineStage):
    """采样阶段 - 对数据进行采样"""
    
    def __init__(self):
        super().__init__("sample")
    
    def execute(self, data: List[Dict[str, Any]], context: PipelineContext) -> List[Dict[str, Any]]:
        if context.sampling_strategy is None:
            return data
        
        try:
            if context.sampling_strategy == "smart":
                return self._smart_sample(data, context)
            elif context.sampling_strategy == "scatter":
                return self._scatter_sample(data, context)
            elif context.sampling_strategy == "byte_limit":
                return self._byte_limit_sample(data, context)
            elif context.sampling_strategy == "random":
                return self._random_sample(data, context)
            else:
                return data
        except Exception as e:
            logger.warning(f"Sampling stage failed: {e}, returning original data")
            return data
    
    def _smart_sample(self, data: List[Dict[str, Any]], context: PipelineContext) -> List[Dict[str, Any]]:
        """智能采样 - 保留数据分布特征"""
        from pygwalker.utils.estimate_tools import smart_sample_datas
        
        if context.sample_size is None:
            return data
        
        return smart_sample_datas(
            data, 
            context.sample_size, 
            preserve_boundaries=context.preserve_boundaries
        )
    
    def _scatter_sample(self, data: List[Dict[str, Any]], context: PipelineContext) -> List[Dict[str, Any]]:
        """散点图专用采样"""
        from pygwalker._constants import (
            SCATTER_PLOT_SAMPLE_LIMIT,
            SCATTER_PLOT_LARGE_DATA_THRESHOLD
        )
        from pygwalker.utils.estimate_tools import smart_sample_datas
        
        n = len(data)
        
        if n <= SCATTER_PLOT_LARGE_DATA_THRESHOLD:
            return data
        
        sample_size = min(SCATTER_PLOT_SAMPLE_LIMIT, SCATTER_PLOT_LARGE_DATA_THRESHOLD // 2)
        
        return smart_sample_datas(data, sample_size, preserve_boundaries=True)
    
    def _byte_limit_sample(self, data: List[Dict[str, Any]], context: PipelineContext) -> List[Dict[str, Any]]:
        """基于字节限制的采样"""
        from pygwalker.utils.estimate_tools import sample_datas_by_byte_limit
        
        if context.byte_limit is None:
            return data
        
        return sample_datas_by_byte_limit(
            data, 
            context.byte_limit,
            min_sample_size=context.sample_size or 100
        )
    
    def _random_sample(self, data: List[Dict[str, Any]], context: PipelineContext) -> List[Dict[str, Any]]:
        """随机采样"""
        import random
        
        if context.sample_size is None or len(data) <= context.sample_size:
            return data
        
        indices = sorted(random.sample(range(len(data)), context.sample_size))
        return [data[i] for i in indices]


class DataPipeline:
    """统一数据管道 - 协调各个阶段的执行"""
    
    def __init__(
        self,
        data_source: DataSource,
        stages: Optional[List[PipelineStage]] = None,
        context: Optional[PipelineContext] = None
    ):
        self.data_source = data_source
        self.stages = stages or [
            FilterStage(),
            SamplingStage()
        ]
        self.context = context or PipelineContext(source_type="unknown")
    
    @classmethod
    def create(
        cls,
        data_source: DataSource,
        **context_kwargs
    ) -> "DataPipeline":
        """工厂方法创建数据管道"""
        context = PipelineContext(
            source_type=getattr(data_source, 'dataset_type', 'unknown'),
            **context_kwargs
        )
        return cls(data_source, context=context)
    
    def configure(self, **kwargs) -> "DataPipeline":
        """配置管道上下文"""
        self.context.update(**kwargs)
        return self
    
    def add_stage(self, stage: PipelineStage) -> "DataPipeline":
        """添加处理阶段"""
        self.stages.append(stage)
        return self
    
    def _fetch_data(self) -> List[Dict[str, Any]]:
        """从数据源获取原始数据"""
        if self.context.payload is not None:
            return self.data_source.get_datas_by_payload(self.context.payload)
        elif self.context.sql is not None:
            return self.data_source.get_datas_by_sql(self.context.sql)
        else:
            raise ValueError("Either payload or sql must be provided in context")
    
    def execute(
        self,
        payload: Optional[Dict[str, Any]] = None,
        sql: Optional[str] = None,
        **kwargs
    ) -> List[Dict[str, Any]]:
        """
        执行数据管道
        
        Args:
            payload: 数据查询 payload
            sql: 数据查询 SQL
            **kwargs: 其他上下文参数
            
        Returns:
            处理后的数据列表
        """
        if payload is not None:
            self.context.payload = payload
        if sql is not None:
            self.context.sql = sql
        
        if kwargs:
            self.context.update(**kwargs)
        
        data = self._fetch_data()
        
        for stage in self.stages:
            data = stage.execute(data, self.context)
        
        return data
    
    def execute_batch(
        self,
        payload_list: Optional[List[Dict[str, Any]]] = None,
        sql_list: Optional[List[str]] = None,
        **kwargs
    ) -> List[List[Dict[str, Any]]]:
        """
        批量执行数据管道
        
        Args:
            payload_list: 数据查询 payload 列表
            sql_list: 数据查询 SQL 列表
            **kwargs: 其他上下文参数
            
        Returns:
            处理后的数据列表的列表
        """
        if kwargs:
            self.context.update(**kwargs)
        
        if payload_list is not None:
            raw_datas = self.data_source.batch_get_datas_by_payload(payload_list)
        elif sql_list is not None:
            raw_datas = self.data_source.batch_get_datas_by_sql(sql_list)
        else:
            raise ValueError("Either payload_list or sql_list must be provided")
        
        results = []
        for data in raw_datas:
            for stage in self.stages:
                data = stage.execute(data, self.context)
            results.append(data)
        
        return results


class PipelineBuilder:
    """数据管道构建器"""
    
    def __init__(self, data_source: DataSource):
        self.data_source = data_source
        self.stages: List[PipelineStage] = []
        self.context_kwargs: Dict[str, Any] = {}
    
    def with_filter(
        self,
        filters: List[Dict[str, Any]],
        logic: str = "AND"
    ) -> "PipelineBuilder":
        """添加过滤配置"""
        if filters:
            self.stages.append(FilterStage())
            self.context_kwargs["filters"] = filters
            self.context_kwargs["filter_logic"] = logic
        return self
    
    def with_sampling(
        self,
        strategy: str,
        sample_size: Optional[int] = None,
        byte_limit: Optional[int] = None,
        preserve_boundaries: bool = False
    ) -> "PipelineBuilder":
        """添加采样配置"""
        self.stages.append(SamplingStage())
        self.context_kwargs["sampling_strategy"] = strategy
        if sample_size is not None:
            self.context_kwargs["sample_size"] = sample_size
        if byte_limit is not None:
            self.context_kwargs["byte_limit"] = byte_limit
        self.context_kwargs["preserve_boundaries"] = preserve_boundaries
        return self
    
    def with_scatter_sampling(self) -> "PipelineBuilder":
        """添加散点图专用采样"""
        return self.with_sampling("scatter")
    
    def with_byte_limit_sampling(self, byte_limit: int) -> "PipelineBuilder":
        """添加基于字节限制的采样"""
        return self.with_sampling("byte_limit", byte_limit=byte_limit)
    
    def with_smart_sampling(self, sample_size: int, preserve_boundaries: bool = True) -> "PipelineBuilder":
        """添加智能采样"""
        return self.with_sampling(
            "smart", 
            sample_size=sample_size, 
            preserve_boundaries=preserve_boundaries
        )
    
    def with_metadata(self, **metadata) -> "PipelineBuilder":
        """添加元数据"""
        existing_metadata = self.context_kwargs.get("metadata", {})
        existing_metadata.update(metadata)
        self.context_kwargs["metadata"] = existing_metadata
        return self
    
    def build(self) -> DataPipeline:
        """构建数据管道"""
        return DataPipeline(
            self.data_source,
            stages=self.stages,
            context=PipelineContext(
                source_type=getattr(self.data_source, 'dataset_type', 'unknown'),
                **self.context_kwargs
            )
        )


def create_pipeline_builder(data_source: DataSource) -> PipelineBuilder:
    """创建管道构建器的便捷函数"""
    return PipelineBuilder(data_source)
